### PR TITLE
Sword Reflection

### PR DESCRIPTION
**Pull request by LordFowl**
Energy swords, shields, and ninja blades can now deflect projectiles.

Energy swords have a fairly moderate chance of reflecting energy weapons.
Energy shields have a much higher chance of reflecting energy weapons.
The ninja blade has the highest chance of reflecting energy weapons.

Energy sword and ninja blade can only effectively reflect shots to the groin, chest, and head.
Energy shield has full body coverage.

Energy sword has less "durability" than shield and blade (Can withstand less shots before shutting off.)

Neither sword, shield, nor blade can deflect shots to the back. Only the ablative vest provides full coverage.

Probably some other stuff I'm forgetting.

### CODE REVIEW

**Comment by skull132** [comment]
Merge conflict from #1612 probably?

**Comment by NanakoAC** [comment]
swords deflecting blasts seems dumb. The ninja one is sort of believable, perhaps due to some guidance system built into the suit. But a random energy sword being able to deflect lasers at any appreciable probability is dumb.

If it did happen it'd be rare and blind luck, like 3% chance. much more than that is nonsensical

File: code/game/objects/items/weapons/melee/energy.dm [hunk 1]
```diff
@@ -8,6 +8,10 @@
 	armor_penetration = 50
 	flags = NOBLOODY
 	can_embed = 0//No embedding pls
+	var/base_reflectchance = 40
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/melee/energy.dm]: 40% is absurd. make this 10% or less pls
energy swords were already crazy powerful before

File: code/game/objects/items/weapons/melee/energy.dm [hunk 2]
```diff
@@ -103,6 +194,11 @@
 	attack_verb = list("attacked", "chopped", "cleaved", "torn", "cut")
 	sharp = 1
 	edge = 1
+	base_reflectchance = 0
+	base_block_chance = 0 //cannot be used to block guns
+	shield_power = 0
+	can_block_bullets = 0
+	armor_penetration = 80
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/melee/energy.dm]: these armour penetration values are insane too. 80 is the highest melee armoour you can usually get. This entirely negates armor. You're removing all reason to even bother with riot armor

Theres no cause for armor pen values to be this high. Even a value as low as 15 is pretty powerful, 30 is pushing it, that much essentially cripples armor. 

This is basically removing a mechanic completely. have you no sense of balance?

**Comment by NanakoAC** [comment]
Some math for you.

Riot armor, with its 80 melee armor, works out to a 72% reduction in incoming damage. And that translates to a total of 3.57x your normal survival time/number of survivable hits. It is obviously a huge advantage

Now with an armor penetration of only 10, thus counting the armor value at 70, the wearer gets a 59.5% reduction in damage, which translates to 2.47x survival time.

Do you see the colossal difference here from just a tiny amount?  That ten points of armor penetration reduces the armor-wearer's survivability (and thus time/number of hits required to kill them) by 30%. 

An armor penetration value of 30, which is the highest i'd consider reasonable, drops the armor value from 80 to 50. And at 50 armor value the effective damage reduction is 37.5%, resulting in a survival time of 1.6x

This means that an armor pen of 30 reduces the survival time by 55%, effectively halving the time and hits required to kill.

You are dealing with an extremely powerful mechanic here, you need to understand the math before throwing random numbers out. Our armor system is a sensitive beast.

I've explained the math behind armor in a post months ago: https://github.com/Aurorastation/Aurora.3/pull/585#issuecomment-233107177

there you can find a list of sample values and a formula to calculate it. The takeaway from that though, should be that the armor system scales up slowly and accelerates at high values. Very low values are almost worthless, and the higher an armor value already is, the bigger an effect armor penetration has on it.

What needs to be avoided is making armor penetration so powerful that armor itself is pointless, and thats a line you've crossed here. At that point you've taken away the advantage that security has over a normal crewmember, and thus taken away a major deterrent to validhunting.

TL:DR, no armor pen values over 30, drop most of them to 5-15 armor pen, except the absolute strongest things like the glaive.

**Comment by NanakoAC** [comment]
Oh and, last but not least, sprinting. The addition of sprinting has made closing gaps far faster, ameliorating the main disadvantage of range that melee weapons have. Please factor this in too. High powered melee weapons are pretty OP

**Comment by NanakoAC** [comment]
As per discussion in discord, these AP values are proportional to the pre-existing ones added by baymerge, so the issue isnt specifically with this PR. We agreed that ill go and fix it seperately.

So, changes approved for now, but ping me when this is merged so i can do a follow up balance pass on melee weapons with AP

File: code/game/objects/items/weapons/melee/energy.dm [hunk 3]
```diff
@@ -51,6 +55,88 @@
 	add_fingerprint(user)
 	return
 
+/obj/item/weapon/melee/energy/handle_shield(mob/user, var/damage, atom/damage_source = null, mob/attacker = null, var/def_zone = null, var/attack_text = "the attack")
+	if(active && default_parry_check(user, attacker, damage_source) && prob(50))
+		user.visible_message("<span class='danger'>\The [user] parries [attack_text] with \the [src]!</span>")
+
+//		Disabled because lag. Immense amounts of lag.
+//		var/datum/effect/effect/system/spark_spread/spark_system = new /datum/effect/effect/system/spark_spread()
```
Comment (NanakoAC) [inline code/game/objects/items/weapons/melee/energy.dm]: lohikar has fixed spark lag now, add this part back in

Comment (Lohikar) [inline code/game/objects/items/weapons/melee/energy.dm]: I think this is actually a merge conflict, I added it back in in my sparks fix PR.

File: code/game/objects/items/weapons/melee/energy.dm [hunk 4]
```diff
@@ -51,6 +55,88 @@
...
+		if(user.incapacitated())
+			return 0
+
+		//block as long as they are not directly behind us
+		var/bad_arc = reverse_direction(user.dir) //arc of directions from which we cannot block
+		if(check_shield_arc(user, bad_arc, damage_source, attacker))
+
+			if(prob(base_block_chance))
+				var/datum/effect/effect/system/spark_spread/spark_system = getFromPool(/datum/effect/effect/system/spark_spread)
+				spark_system.set_up(5, 0, user.loc)
```
Comment (Lohikar) [inline code/game/objects/items/weapons/melee/energy.dm]: This must be updated to use the new spark calls, the old ones have been removed.

File: code/game/objects/items/weapons/shields.dm [hunk 5]
```diff
@@ -103,18 +103,59 @@
...
+	if(user.incapacitated())
+		return 0
+
+	//block as long as they are not directly behind us
+	var/bad_arc = reverse_direction(user.dir) //arc of directions from which we cannot block
+	if(check_shield_arc(user, bad_arc, damage_source, attacker))
+
+		if(prob(get_block_chance(user, damage, damage_source, attacker)))
+			var/datum/effect/effect/system/spark_spread/spark_system = getFromPool(/datum/effect/effect/system/spark_spread)
+			spark_system.set_up(5, 0, user.loc)
```
Comment (Lohikar) [inline code/game/objects/items/weapons/shields.dm]: Needs updating.

File: code/game/objects/items/weapons/melee/energy.dm (hunk 4 above)
Comment (Lohikar) [inline code/game/objects/items/weapons/melee/energy.dm]: For reference: https://github.com/Aurorastation/Aurora.3/blob/development/code/modules/effects/sparks/_spark_docs.dm

File: code/game/objects/items/weapons/melee/energy.dm (hunk 1 above)
Comment (LordFowl) [inline code/game/objects/items/weapons/melee/energy.dm]: This gets subtracted by the projectile's damage divided by 3, so a laser that does 30 damage would have a reflectchance of 30, and so on. The more powerful your laser the less likely it reflects.

File: code/game/objects/items/weapons/melee/energy.dm (hunk 2 above)
Comment (LordFowl) [inline code/game/objects/items/weapons/melee/energy.dm]: Fuck off with those "have you no sense of balance". Completely unnecessary.

**Comment by LordFowl** [comment]
Regular e-swords will no longer even block bullets. Only the ninja sword can block bullets, and it does this with the same regularity as it could deflect bullets in ages past.
